src/groth16/aggregate/prove.rs

```diff
@@ -61,86 +58,81 @@ pub fn aggregate_proofs<E: Engine + std::fmt::Debug>(
         counter_nonce += 1;
     };
 
+    // r, r^2, r^3, r^4 ...
     let r_vec = structured_scalar_power(proofs.len(), &r);
+    // r^-1, r^-2, r^-3
+    let r_inv = r_vec
+        .par_iter()
+        .map(|ri| ri.inverse().unwrap())
+        .collect::<Vec<_>>();
 
-    let r_vec = &r_vec;
-    let ck_1 = &ck_1;
-    let a = &a;
-
-    par! {
-        let a_r = a.par_iter()
-                   .zip(r_vec.par_iter())
-                   .map(|(a, r)| mul!(a.into_projective(), *r).into_affine())
-                   .collect::<Vec<E::G1Affine>>(),
-        let ck_1_r = ck_1.par_iter()
-                         .zip(r_vec.par_iter())
-                         .map(|(ck, r)| mul!(ck.into_projective(), r.inverse().unwrap()).into_affine())
-                         .collect::<Vec<E::G2Affine>>()
-    };
-
-    let a_r = &a_r;
-    let ck_1_r = &ck_1_r;
-    let b = &b;
-    let c = &c;
-
-    par! {
-        let computed_com_a = inner_product::pairing::<E>(a_r, ck_1_r),
-        let tipa_proof_ab = prove_with_srs_shift::<E>(
-                &ip_srs,
-                (a_r, b),
-                (ck_1_r, &ck_2),
-                &r,
-        ),
-        let tipa_proof_c = prove_with_structured_scalar_message::<E>(
-            &ip_srs,
-            (c, r_vec),
-            &ck_1,
-        ),
-        let ip_ab = inner_product::pairing::<E>(&a_r, b),
-        let agg_c = inner_product::multiexponentiation::<E::G1Affine>(&c, r_vec)
-    };
-
-    assert_eq!(com_a, computed_com_a);
+    // A^{r}
+    let a_r = a
+        .par_iter()
+        .zip(r_vec.par_iter())
+        .map(|(ai, ri)| mul!(ai.into_projective(), ri.into_repr()).into_affine())
+        .collect::<Vec<_>>();
+    // V^{r^{-1}}
+    let vkey_r_inv = vkey.scale(&r_inv);
+
+    let tipa_proof_ab = prove_tipp::<E>(&ip_srs, &a_r, &b, &vkey_r_inv, &wkey, &r);
+    let tipa_proof_c = prove_mipp::<E>(
+        &ip_srs, &c, &r_vec,
+        // v - note we dont use the rescaled here since we dont need the
+        // trick as in AB - we just need to commit to C normally.
+        &vkey,
+    );
+    let ip_ab = inner_product::pairing::<E>(&a_r, &b);
+    let agg_c = inner_product::multiexponentiation::<E::G1Affine>(&c, &r_vec);
+
+    // TODO - move assertion to a test - this is a property of the scheme
```

Comment:
Could be `debug_assert`

Reply:
This is taking more time for the prover to compute this `computed_com_ab` - as long s he's doing the computation, it doesn't change for me what we do with it ?

Reply:
if you write it such that the computation is in the `debug_asser_eq!` macro it will get compiled out